Crashes on enum with full comment

```c
/**
 * Comment
 */
typedef enum foo {
    bar = 1
} foo;

int main() {
}
```

Problematic piece of AST:

```
|-EnumDecl 0x5fd8be673708 <1.c:4:9, line:6:1> line:4:14 foo
| |-FullComment 0x5fd8be673ae0 <line:2:3, col:10>
| | `-ParagraphComment 0x5fd8be673ab0 <col:3, col:10>
| |   `-TextComment 0x5fd8be673a80 <col:3, col:10> Text=" Comment"
| `-EnumConstantDecl 0x5fd8be6737e0 <line:5:5, col:11> col:5 bar 'int'
|   `-IntegerLiteral 0x5fd8be6737c0 <col:11> 'int' 1
```

Error:
```
panic: interface conversion: ast.Node is *ast.FullComment, not *ast.EnumConstantDecl

goroutine 1 [running]:
github.com/elliotchance/c2go/transpiler.transpileEnumDecl(0xc4200fe000, 0xc4200da500, 0xc420043490, 0xc4200fe000, 0x6dd920, 0xc4200da500, 0x0)
	c2go/transpiler/enum.go:222 +0x20d8
github.com/elliotchance/c2go/transpiler.transpileToNode(0x6dd920, 0xc4200da500, 0xc4200fe000, 0x0, 0x0, 0x0, 0x0, 0x0)
	c2go/transpiler/transpiler.go:487 +0x1fc
```

Fix:
```diff
diff --git a/transpiler/enum.go b/transpiler/enum.go
index 800182a..f812445 100644
--- a/transpiler/enum.go
+++ b/transpiler/enum.go
@@ -219,7 +219,11 @@ func transpileEnumDecl(p *program.Program, n *ast.EnumDecl) (decls []goast.Decl,
        // counter for replace iota
        var counter int
        for i, c := range n.Children() {
-               e, newPre, newPost := transpileEnumConstantDecl(p, c.(*ast.EnumConstantDecl))
+               cd, ok := c.(*ast.EnumConstantDecl)
+               if !ok {
+                       continue
+               }
+               e, newPre, newPost := transpileEnumConstantDecl(p, cd)
                preStmts, postStmts = combinePreAndPostStmts(preStmts, postStmts, newPre, newPost)
 
                e.Names[0].Obj = goast.NewObj(goast.Con, e.Names[0].Name)
```

--- a/tests/enum.c
+++ b/tests/enum.c
@@ -1,19 +1,27 @@
 #include <stdio.h>
 #include "tests.h"
 
+/* Text */
 enum number{zero, one, two, three};
 
+/**
+ * Text
+ */
 enum
 {
   _ISupper = ((0) < 8 ? ((1 << (0)) << 8) : ((1 << (0)) >> 8)),
   _ISalnum = ((11) < 8 ? ((1 << (11)) << 8) : ((1 << (11)) >> 8))
 };
 
+/** <b> Text </b> */
 enum year{Jan, Feb, Mar, Apr, May, Jun, Jul, 
           Aug, Sep, Oct, Nov, Dec};
 
+// Text
 enum State {Working = 1, Failed = 0, Freezed = 0};
 
+// Text
+// Text
 enum day {sunday = 1, monday, tuesday = 5,
           wednesday, thursday = 10, friday, saturday}; 
 
@@ -22,12 +30,15 @@ enum state currState = 2;
 enum state FindState() {return currState;}
 
 enum { FLY , JUMP };
-// TYPEDEF
 
+/// TYPEDEF
 typedef enum {
     a, b, c
  } T_ENUM;
 
+/**
+ * Text 
+ */
 typedef enum e_strategy {RANDOM, IMMEDIATE = 5, SEARCH} strategy;
 
 // main function

--- a/tests/struct.c
+++ b/tests/struct.c
@@ -64,7 +64,7 @@ int summator(int i, float f){
 
 int main()
 {
-    plan(44);
+    plan(46);
 
     struct programming variable;
     char *s = "Programming in Software Development.";
@@ -228,6 +228,24 @@ int main()
 		f = summator;
 		is_eq(f(3,5),8);
 	}
+
+	diag("typedef struct C C inside function")
+	{
+		typedef struct CCC CCC;
+		struct CCC {
+			float ff;
+		};
+		CCC c;
+		c.ff = 3.14;
+		is_eq(c.ff,3.14);
+	}
+	typedef struct CP CP;
+	struct CP {
+		float ff;
+	};
+	CP cp;
+	cp.ff = 3.14;
+	is_eq(cp.ff,3.14);
 
 	diag("struct name from Go keyword")
 	{ struct chan        {int i;}; struct chan        UU;  UU.i = 5; is_eq(UU.i,5);}

--- a/tests/struct.c
+++ b/tests/struct.c
@@ -62,9 +62,34 @@ int summator(int i, float f){
 	return i+(int)(f);
 }
 
+typedef struct J J;
+struct J
+{
+	float f;
+	int (*fu)(J *j, float i);
+};
+
+int j_function(J *j, float i)
+{
+	if (j != NULL)
+	{
+		return (int)(i+(*j).f);
+	}
+	return -1;
+};
+
+void struct_with_rec_fuction()
+{
+	J j;
+	j.f = 5.0;
+	j.fu = j_function;
+	is_eq(j.fu(&j,4.0),9);
+	is_eq(j_function(NULL, 4.0),-1);
+}
+
 int main()
 {
-    plan(46);
+    plan(48);
 
     struct programming variable;
     char *s = "Programming in Software Development.";
@@ -284,6 +309,8 @@ int main()
 	{ typedef struct {int i;} _           ;	_           UU; UU.i = 5; is_eq(UU.i,5);}
 	{ typedef struct {int i;} init        ;	init        UU; UU.i = 5; is_eq(UU.i,5);}
 */
+
+	struct_with_rec_fuction();
 
     done_testing();
 }

--- a/tests/struct.c
+++ b/tests/struct.c
@@ -29,18 +29,27 @@ void pass_by_val(struct programming value)
     is_streq(value.pointer, "Programming in Software Development.");
 }
 
+/**
+ * text
+ */
 typedef struct mainStruct{
     double constant;
 } secondStruct;
 
+/*
+ * Text
+ */
 typedef struct {
     double t;
 } ts_c;
 
+// Text
 typedef struct ff {
     int v1,v2;
 } tt1, tt2;
 
+// Text1
+// Text2
 struct outer {
     int i;
     struct z {
@@ -50,6 +59,9 @@ struct outer {
 
 struct xx {
     int i;
+	/**
+	 * Text
+	 */
     struct yy {
         int j;
         struct zz {
@@ -58,6 +70,9 @@ struct xx {
     } inner;
 };
 
+/**
+ * Some function
+ */
 int summator(int i, float f){
 	return i+(int)(f);
 }

--- a/tests/union.c
+++ b/tests/union.c
@@ -78,9 +78,46 @@ void union_inside_struct()
 	is_eq(sha2.ff2 , 12.444);
 }
 
+typedef union myunion myunion;
+typedef union myunion
+{
+	double PI;
+	int B;
+}MYUNION;
+
+typedef union
+{
+	double PI;
+	int B;
+}MYUNION2;
+
+void union_typedef()
+{
+	diag("Typedef union")
+	union myunion m;
+	m.PI = 3.14;
+	is_eq(m.PI,3.14);
+	is_true(m.B != 0);
+
+	MYUNION mm;
+	mm.PI = 3.14;
+	is_eq(mm.PI,3.14);
+	is_true(mm.B != 0);
+
+	myunion mmm;
+	mmm.PI = 3.14;
+	is_eq(mmm.PI,3.14);
+	is_true(mmm.B != 0);
+
+	MYUNION2 mmmm;
+	mmmm.PI = 3.14;
+	is_eq(mmmm.PI,3.14);
+	is_true(mmmm.B != 0);
+}
+
 int main()
 {
-    plan(11);
+    plan(19);
 
     union programming variable;
 
@@ -89,6 +126,7 @@ int main()
     pass_by_ref(&variable);
 
 	union_inside_struct();
+	union_typedef();
 
     done_testing();
 }

--- a/tests/union.c
+++ b/tests/union.c
@@ -95,9 +95,14 @@ void union_typedef()
 {
 	diag("Typedef union")
 	union myunion m;
-	m.PI = 3.14;
+	double v = 3.14;
+	m.PI = v;
 	is_eq(m.PI,3.14);
 	is_true(m.B != 0);
+	is_eq(v, 3.14);
+	v += 1.0;
+	is_eq(v, 4.14);
+	is_eq(m.PI,3.14);
 
 	MYUNION mm;
 	mm.PI = 3.14;
@@ -162,9 +167,27 @@ void union_pointers()
 	pass("ok")
 }
 
+union UPNTF{
+	int (*f1)(int);
+	int (*f2)(int);
+};
+
+int union_function(int a)
+{
+	return a+1;
+}
+
+void union_func_pointers()
+{
+	union UPNTF u;
+	u.f1 = union_function;
+	is_eq(u.f1(21), 22);
+	is_eq(u.f2(21), 22);
+}
+
 int main()
 {
-    plan(28);
+    plan(33);
 
     union programming variable;
 
@@ -176,6 +199,7 @@ int main()
 	union_typedef();
 	union_inside_struct2();
 	union_pointers();
+	union_func_pointers();
 
     done_testing();
 }

--- a/tests/union.c
+++ b/tests/union.c
@@ -115,9 +115,56 @@ void union_typedef()
 	is_true(mmmm.B != 0);
 }
 
+typedef struct FuncDestructor FuncDestructor;
+struct FuncDestructor {
+	int i;
+};
+typedef struct FuncDef FuncDef;
+struct FuncDef {
+  int i;
+  union {
+    FuncDef *pHash;
+    FuncDestructor *pDestructor;
+  } u;
+};
+
+void union_inside_struct2()
+{
+	FuncDef f;
+	FuncDestructor fd;
+	fd.i = 100;
+	f.u.pDestructor = &fd;
+
+	FuncDestructor * p_fd = f.u.pDestructor;
+	is_eq((*p_fd).i , 100);
+
+	is_true(f.u.pHash       != NULL);
+	is_true(f.u.pDestructor != NULL);
+	int vHash = (*f.u.pHash).i;
+	is_eq(vHash          , 100);
+	is_eq((*f.u.pHash).i , 100);
+}
+
+union UPNT{
+	int * a;
+	int * b;
+	int * c;
+};
+
+void union_pointers()
+{
+	union UPNT u;
+	int v = 32;
+	u.a = &v;
+	is_eq(*u.a,32);
+	is_eq(*u.b,32);
+	is_eq(*u.c,32);
+	pass("ok")
+}
+
 int main()
 {
-    plan(19);
+    plan(28);
 
     union programming variable;
 
@@ -127,6 +174,8 @@ int main()
 
 	union_inside_struct();
 	union_typedef();
+	union_inside_struct2();
+	union_pointers();
 
     done_testing();
 }

--- a/transpiler/binary.go
+++ b/transpiler/binary.go
@@ -296,6 +296,9 @@ func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsSt
 				ref := memberExpr.GetDeclRefExpr()
 				if ref != nil {
 					union := p.GetStruct(ref.Type)
+					if union == nil {
+						union = p.GetStruct("union " + ref.Type)
+					}
 					if union != nil && union.IsUnion {
 						attrType, err := types.ResolveType(p, ref.Type)
 						if err != nil {

--- a/transpiler/binary.go
+++ b/transpiler/binary.go
@@ -291,6 +291,26 @@ func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsSt
 				right = util.NewNil()
 			}
 
+			var haveMemberExprWithUnion bool
+			var findUnion func(ast.Node)
+			var UnionNode ast.Node
+			findUnion = func(node ast.Node) {
+				switch chi := node.(type) {
+				case *ast.MemberExpr:
+					if strings.HasPrefix(chi.Type, "union ") {
+						haveMemberExprWithUnion = true
+						UnionNode = node
+					}
+					findUnion(node.Children()[0])
+				case *ast.DeclRefExpr:
+					if strings.HasPrefix(chi.Type, "union ") {
+						haveMemberExprWithUnion = true
+						UnionNode = node
+					}
+				}
+			}
+			findUnion(n.Children()[0])
+
 			// Construct code for assigning value to an union field
 			if memberExpr, ok := n.Children()[0].(*ast.MemberExpr); ok {
 				ref := memberExpr.GetDeclRefExpr()
@@ -330,6 +350,12 @@ func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsSt
 							return resExpr, resType, preStmts, postStmts, nil
 						}
 					}
+				}
+
+				// struct inside union
+				if haveMemberExprWithUnion {
+					p.AddMessage(p.GenerateWarningMessage(
+						fmt.Errorf("Binary operation with union not support. AST node with union : %v", UnionNode), n))
 				}
 			}
 		}

--- a/transpiler/binary.go
+++ b/transpiler/binary.go
@@ -319,11 +319,7 @@ func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsSt
 					union := p.GetStruct(un.Type)
 					if union != nil && union.IsUnion {
 						if str, ok := un.Children()[0].(*ast.DeclRefExpr); ok {
-							attrType, err := types.ResolveType(p, memberExpr.Type)
-							if err != nil {
-								p.AddMessage(p.GenerateWarningMessage(err, memberExpr))
-							}
-							funcName := getFunctionNameForUnionSetter("", attrType, memberExpr.Name)
+							funcName := getFunctionNameForUnionSetter("", memberExpr.Type, memberExpr.Name)
 							funcName = str.Name + "." + un.Name + funcName
 							resExpr := &goast.CallExpr{
 								Fun:  goast.NewIdent(funcName),

--- a/transpiler/binary.go
+++ b/transpiler/binary.go
@@ -52,6 +52,25 @@ func transpileBinaryOperatorComma(n *ast.BinaryOperator, p *program.Program) (
 	return right[len(right)-1], preStmts, nil
 }
 
+func findUnion(node ast.Node) (unionNode ast.Node, haveMemberExprWithUnion bool) {
+	switch chi := node.(type) {
+	case *ast.MemberExpr:
+		if strings.HasPrefix(chi.Type, "union ") {
+			haveMemberExprWithUnion = true
+			unionNode = node
+			return
+		}
+		return findUnion(node.Children()[0])
+	case *ast.DeclRefExpr:
+		if strings.HasPrefix(chi.Type, "union ") {
+			haveMemberExprWithUnion = true
+			unionNode = node
+			return
+		}
+	}
+	return nil, false
+}
+
 func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsStmt bool) (
 	_ goast.Expr, resultType string, preStmts []goast.Stmt, postStmts []goast.Stmt, err error) {
 	defer func() {
@@ -271,25 +290,7 @@ func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsSt
 				right = util.NewNil()
 			}
 
-			var haveMemberExprWithUnion bool
-			var findUnion func(ast.Node)
-			var UnionNode ast.Node
-			findUnion = func(node ast.Node) {
-				switch chi := node.(type) {
-				case *ast.MemberExpr:
-					if strings.HasPrefix(chi.Type, "union ") {
-						haveMemberExprWithUnion = true
-						UnionNode = node
-					}
-					findUnion(node.Children()[0])
-				case *ast.DeclRefExpr:
-					if strings.HasPrefix(chi.Type, "union ") {
-						haveMemberExprWithUnion = true
-						UnionNode = node
-					}
-				}
-			}
-			findUnion(n.Children()[0])
+			unionNode, haveMemberExprWithUnion := findUnion(n.Children()[0])
 
 			// Construct code for assigning value to an union field
 			if memberExpr, ok := n.Children()[0].(*ast.MemberExpr); ok {
@@ -335,7 +336,7 @@ func transpileBinaryOperator(n *ast.BinaryOperator, p *program.Program, exprIsSt
 				// struct inside union
 				if haveMemberExprWithUnion {
 					p.AddMessage(p.GenerateWarningMessage(
-						fmt.Errorf("Binary operation with union not support. AST node with union : %v", UnionNode), n))
+						fmt.Errorf("Binary operation with union not support. AST node with union : %v", unionNode), n))
 				}
 			}
 		}

--- a/transpiler/call.go
+++ b/transpiler/call.go
@@ -24,6 +24,9 @@ func getName(firstChild ast.Node) string {
 	case *ast.MemberExpr:
 		if types.IsFunction(fc.Type) {
 			if decl, ok := fc.Children()[0].(*ast.DeclRefExpr); ok {
+				if strings.Contains(decl.Type, "union ") {
+					return getFunctionNameForUnionGetter(decl.Name, "", fc.Name) + "()"
+				}
 				return decl.Name + "." + fc.Name
 			}
 		}

--- a/transpiler/declarations.go
+++ b/transpiler/declarations.go
@@ -143,11 +143,19 @@ func transpileRecordDecl(p *program.Program, n *ast.RecordDecl) (decls []goast.D
 					pos++
 					decls = append(decls, declUnion...)
 				}
-			} else if field.Kind == "struct" && len(field.Children()) > 0 && field.Name == "" && pos+2 <= len(n.Children()) {
+			} else if field.Kind == "struct" &&
+				len(field.Children()) > 0 &&
+				field.Name == "" &&
+				pos+2 <= len(n.Children()) {
 				if inField, ok := n.Children()[pos+1].(*ast.FieldDecl); ok {
 					inField.Type = types.GenerateCorrectType(inField.Type)
 					inField.Type2 = types.GenerateCorrectType(inField.Type2)
 					field.Name = string(([]byte(inField.Type))[len("struct "):])
+					if strings.Contains(field.Name, "[") {
+						p.AddMessage(p.GenerateWarningMessage(
+							fmt.Errorf("Not acceptable name of struct %s", field.Name), n))
+						continue
+					}
 					declStruct, err := transpileRecordDecl(p, field)
 					if err != nil {
 						p.AddMessage(p.GenerateWarningMessage(err, field))

--- a/transpiler/declarations.go
+++ b/transpiler/declarations.go
@@ -198,7 +198,11 @@ func transpileRecordDecl(p *program.Program, n *ast.RecordDecl) (decls []goast.D
 			p.AddImports("reflect", "unsafe")
 
 			// Declaration for implementing union type
-			decls = append(decls, transpileUnion(name, size, fields)...)
+			d, err2 := transpileUnion(name, size, fields)
+			if err2 != nil {
+				return nil, err2
+			}
+			decls = append(decls, d...)
 		}
 	} else {
 		decls = append(decls, &goast.GenDecl{

--- a/transpiler/declarations.go
+++ b/transpiler/declarations.go
@@ -106,13 +106,6 @@ func transpileRecordDecl(p *program.Program, n *ast.RecordDecl) (decls []goast.D
 	name = types.GenerateCorrectType(name)
 	p.DefineType(name)
 
-	s := program.NewStruct(n)
-	if s.IsUnion {
-		p.Unions["union "+s.Name] = s
-	} else {
-		p.Structs["struct "+s.Name] = s
-	}
-
 	// TODO: Some platform structs are ignored.
 	// https://github.com/elliotchance/c2go/issues/85
 	if name == "__locale_struct" ||
@@ -150,6 +143,18 @@ func transpileRecordDecl(p *program.Program, n *ast.RecordDecl) (decls []goast.D
 					pos++
 					decls = append(decls, declUnion...)
 				}
+			} else if field.Kind == "struct" && len(field.Children()) > 0 && field.Name == "" && pos+2 <= len(n.Children()) {
+				if inField, ok := n.Children()[pos+1].(*ast.FieldDecl); ok {
+					inField.Type = types.GenerateCorrectType(inField.Type)
+					inField.Type2 = types.GenerateCorrectType(inField.Type2)
+					field.Name = string(([]byte(inField.Type))[len("struct "):])
+					declStruct, err := transpileRecordDecl(p, field)
+					if err != nil {
+						p.AddMessage(p.GenerateWarningMessage(err, field))
+					}
+					pos++
+					decls = append(decls, declStruct...)
+				}
 			} else {
 				decls, err = transpileRecordDecl(p, field)
 				if err != nil {
@@ -164,13 +169,19 @@ func transpileRecordDecl(p *program.Program, n *ast.RecordDecl) (decls []goast.D
 		}
 	}
 
+	s := program.NewStruct(n)
+	if s.IsUnion {
+		p.Unions["union "+s.Name] = s
+	} else {
+		p.Structs["struct "+s.Name] = s
+	}
 	if s.IsUnion {
 		// Union size
 		size, err := types.SizeOf(p, "union "+name)
 
 		// In normal case no error is returned,
 		if err != nil {
-			// but if we catch one, send it as a aarning
+			// but if we catch one, send it as a warning
 			message := fmt.Sprintf("could not determine the size of type `union %s` for that reason: %s", name, err)
 			p.AddMessage(p.GenerateWarningMessage(errors.New(message), n))
 		} else {

--- a/transpiler/enum.go
+++ b/transpiler/enum.go
@@ -218,7 +218,12 @@ func transpileEnumDecl(p *program.Program, n *ast.EnumDecl) (decls []goast.Decl,
 
 	// counter for replace iota
 	var counter int
-	for i, c := range n.Children() {
+	var i int
+	for _, c := range n.Children() {
+		if _, ok := c.(*ast.EnumConstantDecl); !ok {
+			// add for avoid comments elements
+			continue
+		}
 		e, newPre, newPost := transpileEnumConstantDecl(p, c.(*ast.EnumConstantDecl))
 		preStmts, postStmts = combinePreAndPostStmts(preStmts, postStmts, newPre, newPost)
 
@@ -274,6 +279,9 @@ func transpileEnumDecl(p *program.Program, n *ast.EnumDecl) (decls []goast.Decl,
 
 		// registration of enum constants
 		p.EnumConstantToEnum[e.Names[0].Name] = "enum " + n.Name
+
+		// calculate next position without comments
+		i++
 	}
 
 	// important value for creating (.....)

--- a/transpiler/translation_unit.go
+++ b/transpiler/translation_unit.go
@@ -85,15 +85,6 @@ func isSameTypedefNames(v *ast.TypedefDecl) bool {
 	     `-RecordType 0x33d9fa0 'struct Uq'
 	       `-Record 0x33da090 'Uq'
 	*/
-	if "struct "+v.Name == v.Type2 && v.Type == v.Type2 {
-		if vv, ok := v.Children()[0].(*ast.ElaboratedType); ok && vv.Type == v.Type {
-			if vvv, ok := vv.Children()[0].(*ast.RecordType); ok && vvv.Type == v.Type2 {
-				if vvvv, ok := vvv.Children()[0].(*ast.Record); ok && vvvv.Type == v.Name {
-					return true
-				}
-			}
-		}
-	}
 	// for unions:
 	/*
 		TypedefDecl 0x38bc070 <col:1, col:23> col:23 referenced myunion 'union myunion':'union myunion'
@@ -101,7 +92,7 @@ func isSameTypedefNames(v *ast.TypedefDecl) bool {
 		  `-RecordType 0x38bc000 'union myunion'
 		    `-Record 0x38bc0d8 'myunion'
 	*/
-	if "union "+v.Name == v.Type2 && v.Type == v.Type2 {
+	if ("struct "+v.Name == v.Type2 || "union "+v.Name == v.Type2) && v.Type == v.Type2 {
 		if vv, ok := v.Children()[0].(*ast.ElaboratedType); ok && vv.Type == v.Type {
 			if vvv, ok := vv.Children()[0].(*ast.RecordType); ok && vvv.Type == v.Type2 {
 				if vvvv, ok := vvv.Children()[0].(*ast.Record); ok && vvvv.Type == v.Name {

--- a/transpiler/translation_unit.go
+++ b/transpiler/translation_unit.go
@@ -2,6 +2,7 @@ package transpiler
 
 import (
 	goast "go/ast"
+	"strings"
 
 	"github.com/elliotchance/c2go/ast"
 	"github.com/elliotchance/c2go/program"
@@ -14,28 +15,18 @@ func transpileTranslationUnitDecl(p *program.Program, n *ast.TranslationUnitDecl
 		case *ast.RecordDecl:
 			// for case :
 			// typedef struct C C;
+			// typedef union  C C;
 			if len(v.Children()) == 0 {
-				strF := v.Name
 				if i+1 < len(n.Children()) {
 					if vv, ok := n.Children()[i+1].(*ast.TypedefDecl); ok {
-						if len(vv.Name)+7 == len(vv.Type) {
-							if "struct "+vv.Name == vv.Type && strF == vv.Name {
-								if vvv, ok := vv.Children()[0].(*ast.ElaboratedType); ok {
-									if vvv, ok := vvv.Children()[0].(*ast.RecordType); ok {
-										if vvv, ok := vvv.Children()[0].(*ast.Record); ok {
-											if vv.Name == vvv.Type {
-												i++
-												break
-											}
-										}
-									}
-								}
-							}
+						if isSameTypedefNames(vv) {
+							i++
+							continue
 						}
 					}
 				}
 			}
-			// specific for `typedef struct` without name
+			// specific for `typedef struct`, `typedef union` without name
 			if v.Name != "" || i == len(n.Children())-1 {
 				var d []goast.Decl
 				d, err = transpileRecordDecl(p, v)
@@ -52,6 +43,10 @@ func transpileTranslationUnitDecl(p *program.Program, n *ast.TranslationUnitDecl
 					// name and fields
 					var recordDecl ast.RecordDecl
 					recordDecl.Name = nameTypedefStruct
+					recordDecl.Kind = "struct"
+					if strings.Contains(vv.Type, "union ") {
+						recordDecl.Kind = "union"
+					}
 					recordDecl.ChildNodes = fields
 					var d []goast.Decl
 					d, err = transpileRecordDecl(p, &recordDecl)
@@ -80,4 +75,40 @@ func transpileTranslationUnitDecl(p *program.Program, n *ast.TranslationUnitDecl
 		}
 	}
 	return
+}
+
+func isSameTypedefNames(v *ast.TypedefDecl) bool {
+	// for structs :
+	/*
+	   TypedefDecl 0x33da010 <col:3, col:21> col:21 referenced Uq 'struct Uq':'struct Uq'
+	   `-ElaboratedType 0x33d9fc0 'struct Uq' sugar
+	     `-RecordType 0x33d9fa0 'struct Uq'
+	       `-Record 0x33da090 'Uq'
+	*/
+	if "struct "+v.Name == v.Type2 && v.Type == v.Type2 {
+		if vv, ok := v.Children()[0].(*ast.ElaboratedType); ok && vv.Type == v.Type {
+			if vvv, ok := vv.Children()[0].(*ast.RecordType); ok && vvv.Type == v.Type2 {
+				if vvvv, ok := vvv.Children()[0].(*ast.Record); ok && vvvv.Type == v.Name {
+					return true
+				}
+			}
+		}
+	}
+	// for unions:
+	/*
+		TypedefDecl 0x38bc070 <col:1, col:23> col:23 referenced myunion 'union myunion':'union myunion'
+		`-ElaboratedType 0x38bc020 'union myunion' sugar
+		  `-RecordType 0x38bc000 'union myunion'
+		    `-Record 0x38bc0d8 'myunion'
+	*/
+	if "union "+v.Name == v.Type2 && v.Type == v.Type2 {
+		if vv, ok := v.Children()[0].(*ast.ElaboratedType); ok && vv.Type == v.Type {
+			if vvv, ok := vv.Children()[0].(*ast.RecordType); ok && vvv.Type == v.Type2 {
+				if vvvv, ok := vvv.Children()[0].(*ast.Record); ok && vvvv.Type == v.Name {
+					return true
+				}
+			}
+		}
+	}
+	return false
 }

--- a/transpiler/unary.go
+++ b/transpiler/unary.go
@@ -183,6 +183,49 @@ func transpilePointerArith(n *ast.UnaryOperator, p *program.Program) (
 				found = true
 				return
 
+			case *ast.MemberExpr:
+				// check - if member of union
+				var ok bool
+				var a ast.Node = n.Children()[i]
+				var b ast.Node
+				var isUnion bool
+				for {
+					if b, ok = a.Children()[0].(*ast.MemberExpr); ok {
+						if strings.Contains(b.(*ast.MemberExpr).Type, "union ") {
+							isUnion = true
+							break
+						}
+						a = b
+						continue
+					} else if b, ok = a.Children()[0].(*ast.DeclRefExpr); ok {
+						if strings.Contains(b.(*ast.DeclRefExpr).Type, "union ") {
+							isUnion = true
+							break
+						}
+						a = b
+						continue
+					}
+					break
+				}
+				if isUnion {
+					counter++
+					if counter > 1 {
+						err = fmt.Errorf("Not acceptable : change counter is more then 1. found = %v,%v", pointer, v)
+						return
+					}
+					// found pointer
+					pointer = v
+					// Replace pointer to zero
+					var zero ast.IntegerLiteral
+					zero.Type = "int"
+					zero.Value = "0"
+					n.Children()[i] = &zero
+					found = true
+					return
+				}
+				// member of struct
+				f(v)
+
 			case *ast.CallExpr:
 				if v.Type == "int" {
 					continue
@@ -264,6 +307,19 @@ func transpilePointerArith(n *ast.UnaryOperator, p *program.Program) (
 	eType = n.Type
 
 	switch v := pointer.(type) {
+	case *ast.MemberExpr:
+		arr, _, newPre, newPost, err2 := transpileToExpr(v, p, false)
+		if err2 != nil {
+			return
+		}
+		preStmts, postStmts = combinePreAndPostStmts(preStmts, postStmts, newPre, newPost)
+		return &goast.IndexExpr{
+			X: &goast.ParenExpr{
+				Lparen: 1,
+				X:      arr,
+			},
+			Index: e,
+		}, eType, preStmts, postStmts, err
 	case *ast.DeclRefExpr:
 		return &goast.IndexExpr{
 			X:     util.NewIdent(v.Name),

--- a/transpiler/union.go
+++ b/transpiler/union.go
@@ -1,293 +1,111 @@
 package transpiler
 
 import (
+	"bytes"
+	"html/template"
 	"strings"
 
 	goast "go/ast"
+	"go/format"
+	"go/parser"
 	"go/token"
 
 	"fmt"
-
-	"github.com/elliotchance/c2go/util"
 )
 
 func transpileUnion(name string, size int, fields []*goast.Field) []goast.Decl {
-	res := []goast.Decl{
-		// Type declaration (array: [x]byte with x the size of union)
-		&goast.GenDecl{
-			Tok: token.TYPE,
-			Specs: []goast.Spec{
-				&goast.TypeSpec{
-					Name: util.NewIdent(name),
-					Type: &goast.ArrayType{
-						Elt: util.NewIdent("byte"),
-						Len: util.NewIntLit(size), // Size of the union
-					},
-				},
-			},
-		},
 
-		// cast() method
-		&goast.FuncDecl{
-			Name: util.NewIdent("cast"),
-			Recv: &goast.FieldList{
-				List: []*goast.Field{
-					{
-						Names: []*goast.Ident{util.NewIdent("self")},
-						Type: &goast.StarExpr{
-							X: util.NewIdent(name),
-						},
-					},
-				},
-			},
-			Type: &goast.FuncType{
-				Params: &goast.FieldList{
-					List: []*goast.Field{
-						{
-							Names: []*goast.Ident{util.NewIdent("t")},
-							Type: &goast.SelectorExpr{
-								X:   util.NewIdent("reflect"),
-								Sel: util.NewIdent("Type"),
-							},
-						},
-					},
-				},
-				Results: &goast.FieldList{
-					List: []*goast.Field{
-						{
-							Type: &goast.SelectorExpr{
-								X:   util.NewIdent("reflect"),
-								Sel: util.NewIdent("Value"),
-							},
-						},
-					},
-				},
-			},
-			Body: &goast.BlockStmt{
-				List: []goast.Stmt{
-					&goast.ReturnStmt{
-						Results: []goast.Expr{
-							util.NewCallExpr("reflect.NewAt",
-								util.NewIdent("t"),
-								&goast.CallExpr{
-									Fun: &goast.SelectorExpr{
-										X:   util.NewIdent("unsafe"),
-										Sel: util.NewIdent("Pointer"),
-									},
-									Args: []goast.Expr{
-										&goast.UnaryExpr{
-											Op: token.AND,
-											X: &goast.IndexExpr{
-												X:     util.NewIdent("self"),
-												Index: util.NewIntLit(0),
-											},
-										},
-									},
-								},
-							),
-						},
-					},
-				},
-			},
-		},
-
-		// assign() method
-		&goast.FuncDecl{
-			Name: util.NewIdent("assign"),
-			Recv: &goast.FieldList{
-				List: []*goast.Field{
-					{
-						Names: []*goast.Ident{util.NewIdent("self")},
-						Type: &goast.StarExpr{
-							X: util.NewIdent(name),
-						},
-					},
-				},
-			},
-			Type: &goast.FuncType{
-				Params: &goast.FieldList{
-					List: []*goast.Field{
-						{
-							Names: []*goast.Ident{util.NewIdent("v")},
-							Type: &goast.InterfaceType{
-								Methods: new(goast.FieldList),
-							},
-						},
-					},
-				},
-			},
-			Body: &goast.BlockStmt{
-				List: []goast.Stmt{
-					&goast.AssignStmt{
-						Lhs: []goast.Expr{
-							util.NewIdent("value"),
-						},
-						Tok: token.DEFINE,
-						Rhs: []goast.Expr{
-							&goast.CallExpr{
-								Fun: &goast.SelectorExpr{
-									X:   util.NewCallExpr("reflect.ValueOf", util.NewIdent("v")),
-									Sel: util.NewIdent("Elem"),
-								},
-							},
-						},
-					},
-					util.NewExprStmt(
-						util.NewCallExpr("value.Set",
-							&goast.CallExpr{
-								Fun: &goast.SelectorExpr{
-									X:   util.NewCallExpr("self.cast", util.NewCallExpr("value.Type")),
-									Sel: util.NewIdent("Elem"),
-								},
-							},
-						),
-					),
-				},
-			},
-		},
-
-		// UntypedSet() method
-		&goast.FuncDecl{
-			Name: util.NewIdent("UntypedSet"),
-			Recv: &goast.FieldList{
-				List: []*goast.Field{
-					{
-						Names: []*goast.Ident{util.NewIdent("self")},
-						Type: &goast.StarExpr{
-							X: util.NewIdent(name),
-						},
-					},
-				},
-			},
-			Type: &goast.FuncType{
-				Params: &goast.FieldList{
-					List: []*goast.Field{
-						{
-							Names: []*goast.Ident{util.NewIdent("v")},
-							Type: &goast.InterfaceType{
-								Methods: new(goast.FieldList),
-							},
-						},
-					},
-				},
-			},
-			Body: &goast.BlockStmt{
-				List: []goast.Stmt{
-					&goast.AssignStmt{
-						Lhs: []goast.Expr{
-							util.NewIdent("value"),
-						},
-						Tok: token.DEFINE,
-						Rhs: []goast.Expr{
-							util.NewCallExpr("reflect.ValueOf", util.NewIdent("v")),
-						},
-					},
-					util.NewExprStmt(
-						&goast.CallExpr{
-							Fun: &goast.SelectorExpr{
-								X: &goast.CallExpr{
-									Fun: &goast.SelectorExpr{
-										X:   util.NewCallExpr("self.cast", util.NewCallExpr("value.Type")),
-										Sel: util.NewIdent("Elem"),
-									},
-								},
-								Sel: util.NewIdent("Set"),
-							},
-							Args: []goast.Expr{
-								util.NewIdent("value"),
-							},
-						},
-					),
-				},
-			},
-		},
+	type field struct {
+		Name      string
+		TypeField string
 	}
 
-	// Methods for each union field
-	for _, f := range fields {
-		fieldID := strings.Title(f.Names[0].Name)
-
-		res = append(res,
-			// Setter method (SetXX)
-			&goast.FuncDecl{
-				Name: util.NewIdent("Set" + fieldID),
-				Recv: &goast.FieldList{
-					List: []*goast.Field{
-						{
-							Names: []*goast.Ident{util.NewIdent("self")},
-							Type: &goast.StarExpr{
-								X: util.NewIdent(name),
-							},
-						},
-					},
-				},
-				Type: &goast.FuncType{
-					Params: &goast.FieldList{
-						List: []*goast.Field{
-							{
-								Names: []*goast.Ident{util.NewIdent("v")},
-								Type:  f.Type,
-							},
-						},
-					},
-					Results: &goast.FieldList{
-						List: []*goast.Field{
-							{
-								Type: f.Type,
-							},
-						},
-					},
-				},
-				Body: &goast.BlockStmt{
-					List: []goast.Stmt{
-						util.NewExprStmt(
-							util.NewCallExpr("self.UntypedSet", util.NewIdent("v")),
-						),
-						&goast.ReturnStmt{
-							Results: []goast.Expr{
-								util.NewIdent("v"),
-							},
-						},
-					},
-				},
-			},
-
-			// Getter method (GetXX)
-			&goast.FuncDecl{
-				Name: util.NewIdent("Get" + fieldID),
-				Recv: &goast.FieldList{
-					List: []*goast.Field{
-						{
-							Names: []*goast.Ident{util.NewIdent("self")},
-							Type: &goast.StarExpr{
-								X: util.NewIdent(name),
-							},
-						},
-					},
-				},
-				Type: &goast.FuncType{
-					Results: &goast.FieldList{
-						List: []*goast.Field{
-							{
-								Names: []*goast.Ident{util.NewIdent("res")},
-								Type:  f.Type,
-							},
-						},
-					},
-				},
-				Body: &goast.BlockStmt{
-					List: []goast.Stmt{
-						util.NewExprStmt(
-							util.NewCallExpr("self.assign", util.NewUnaryExpr(token.AND, util.NewIdent("res"))),
-						),
-						new(goast.ReturnStmt),
-					},
-				},
-			},
-		)
+	type union struct {
+		Name   string
+		Size   int
+		Fields []field
 	}
 
-	return res
+	src := `package main
+
+import(
+	"unsafe"
+	"reflect"
+)
+
+type {{ .Name }} struct{
+	value    interface{}
+	arr      [{{ .Size }}]byte
+}
+
+func (self *{{ .Name }}) cast(t reflect.Type) reflect.Value {
+	return reflect.NewAt(t, unsafe.Pointer(&self.arr[0]))
+}
+
+func (self *{{ .Name }}) assign(v interface{}){
+	value := reflect.ValueOf(v).Elem()
+	value.Set(self.cast(value.Type()).Elem())
+}
+
+func (self *{{ .Name }}) UntypedSet(v interface{}){
+	value := reflect.ValueOf(v)
+	self.cast(value.Type()).Elem().Set(value)
+}
+
+{{ range .Fields }}
+func (self *{{ $.Name }}) Get{{ .Name }} () (res {{ .TypeField }}){
+	self.assign(&res)
+	return
+}
+
+func (self *{{ $.Name }}) Set{{ .Name }} (v {{ .TypeField }}) {{ .TypeField }}{
+	self.value = v // added for avoid GC removing pointers in union
+	self.UntypedSet(v)
+	return v
+}
+{{ end }}
+`
+	// Generate structure of union
+	var un union
+	un.Name = name
+	un.Size = size
+	for i := range fields {
+		var f field
+		f.Name = fields[i].Names[0].Name
+
+		var buf bytes.Buffer
+		err := format.Node(&buf, token.NewFileSet(), fields[i].Type)
+		if err != nil {
+			panic(err)
+		}
+		f.TypeField = buf.String()
+
+		// capitalization first letter
+		name := strings.ToUpper(string(f.Name[0]))
+		if len(f.Name) > 1 {
+			name += f.Name[1:]
+		}
+		f.Name = name
+
+		un.Fields = append(un.Fields, f)
+	}
+
+	var err error
+	tmpl := template.Must(template.New("").Parse(src))
+	var source bytes.Buffer
+	err = tmpl.Execute(&source, un)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create the AST by parsing src.
+	fset := token.NewFileSet() // positions are relative to fset
+	f, err := parser.ParseFile(fset, "", source.String(), 0)
+	if err != nil {
+		panic(err)
+	}
+
+	return f.Decls[1:]
 }
 
 func getFunctionNameForUnion(verb, variableName, variableType, attributeName string) string {

--- a/transpiler/union.go
+++ b/transpiler/union.go
@@ -7,6 +7,7 @@ import (
 	"go/token"
 
 	"fmt"
+
 	"github.com/elliotchance/c2go/util"
 )
 

--- a/transpiler/union.go
+++ b/transpiler/union.go
@@ -13,7 +13,8 @@ import (
 	"fmt"
 )
 
-func transpileUnion(name string, size int, fields []*goast.Field) []goast.Decl {
+func transpileUnion(name string, size int, fields []*goast.Field) (
+	_ []goast.Decl, err error) {
 
 	type field struct {
 		Name      string
@@ -53,11 +54,13 @@ func (self *{{ .Name }}) UntypedSet(v interface{}){
 }
 
 {{ range .Fields }}
+// Get{{ .Name }} - return value of {{ .Name }}
 func (self *{{ $.Name }}) Get{{ .Name }} () (res {{ .TypeField }}){
 	self.assign(&res)
 	return
 }
 
+// Set{{ .Name }} - set value of {{ .Name }}
 func (self *{{ $.Name }}) Set{{ .Name }} (v {{ .TypeField }}) {{ .TypeField }}{
 	self.value = v // added for avoid GC removing pointers in union
 	self.UntypedSet(v)
@@ -90,22 +93,21 @@ func (self *{{ $.Name }}) Set{{ .Name }} (v {{ .TypeField }}) {{ .TypeField }}{
 		un.Fields = append(un.Fields, f)
 	}
 
-	var err error
 	tmpl := template.Must(template.New("").Parse(src))
 	var source bytes.Buffer
 	err = tmpl.Execute(&source, un)
 	if err != nil {
-		panic(err)
+		return
 	}
 
 	// Create the AST by parsing src.
 	fset := token.NewFileSet() // positions are relative to fset
 	f, err := parser.ParseFile(fset, "", source.String(), 0)
 	if err != nil {
-		panic(err)
+		return
 	}
 
-	return f.Decls[1:]
+	return f.Decls[1:], nil
 }
 
 func getFunctionNameForUnion(verb, variableName, variableType, attributeName string) string {

--- a/transpiler/variables.go
+++ b/transpiler/variables.go
@@ -323,6 +323,10 @@ func transpileMemberExpr(n *ast.MemberExpr, p *program.Program) (
 	if structType == nil {
 		structType = p.GetStruct("struct " + lhsType)
 	}
+	// added for support "union typedef"
+	if structType == nil {
+		structType = p.GetStruct("union " + lhsType)
+	}
 	rhs := n.Name
 	rhsType := "void *"
 	if structType == nil {

--- a/util/goast.go
+++ b/util/goast.go
@@ -60,6 +60,21 @@ func internalTypeToExpr(t string) goast.Expr {
 		}
 	}
 
+	// function return a function
+	// Example : u.GetF1()
+	if strings.HasSuffix(t, "()") && strings.Contains(t, ".") {
+		i := strings.Index(t, ".")
+		if i > 0 {
+			t = t[:len(t)-2] // remove "()" at the end
+			return &goast.CallExpr{
+				Fun: &goast.SelectorExpr{
+					X:   NewIdent(t[0:i]),
+					Sel: NewIdent(t[i+1:]),
+				},
+			}
+		}
+	}
+
 	// Parenthesis Expression
 	if strings.HasPrefix(t, "(") && strings.HasSuffix(t, ")") {
 		return &goast.ParenExpr{

